### ISSUE TITLE
False Positive in PL2 with double encoded JSON data

### ISSUE THREAD

**Issue by mirkodziadzka-avi** [coreruleset/coreruleset#2449]
I'm not sure if this is fixable, but I wanted to report it anyway for documentation.

### Description

We have seen multiple clients which seem to double encode json data (e.g. having a serialized JSON document as a value in a JSON message).

Here is a python script which is recreating such a request against the sandbox.

```py
#!/usr/bin/env python3

import json
import requests

URL = "https://sandbox.coreruleset.org/"

message = {
  "id": "hello",
  "value": json.dumps({
    "service": "SomeService",
    "name": "foo",
    "doSomething": False,
    "CustomerNumber": "123456789"
  })
}

headers = {
    "Content-Type": "application/json",
    "X-Backend": "nginx",
    "X-CRS-Paranoia-Level": "2",
    "X-Format-Output": "txt-matched-rules",
}
response = requests.post(URL, headers=headers, data=json.dumps(message))

print(response.text)
```
This request is blocked by multiple SQLi rules in PL2

### Audit Logs / Triggered Rule Numbers

```
942200 PL2 Detects MySQL comment-/space-obfuscated injections and backtick termination
942260 PL2 Detects basic SQL authentication bypass attempts 2/3
942430 PL2 Restricted SQL Character Anomaly Detection (args): # of special characters exceeded (12)
```

### Your Environment

Sandbox.

### Confirmation

[x] I have removed any personal data (email addresses, IP addresses,
    passwords, domain names) from any logs posted.

**Comment by mirkodziadzka-avi** [comment]
cc: @christiantreutler-avi

**Comment by franbuehler** [comment]
Thank you for the reporting this issue.

I can confirm that the JSON data inside the JSON document triggers the following rules (up to PL2):

```
942200 PL2 Detects MySQL comment-/space-obfuscated injections and backtick termination
942260 PL2 Detects basic SQL authentication bypass attempts 2/3
942430 PL2 Restricted SQL Character Anomaly Detection (args): # of special characters exceeded (12)
```



The corresponding audit logs look like this:

```
[Fri Mar 25 20:17:38.349770 2022] [:error] [pid 15:tid 140482004907776] [client 172.x.x.x:58974] [client 172.x.x.x] 
ModSecurity: Warning. Pattern match "(?i)(?:(?:(?:(?:trunc|cre|upd)at|renam)e|d(?:e(?:lete|sc)|rop)|
(?:inser|selec)t|alter|load)\\\\s*?\\\\(\\\\s*?space\\\\s*?\\\\(|,.*?[)\\\\da-f\\"'`][\\"'`](?:[\\"'`].*?[\\"'`]|(?:\\\\r?\\\\n)?\\\\z|[^\\"'`]+)|
\\\\Wselect.+\\\\W*?from)" at ARGS:value. [file "/etc/modsecurity.d/owasp-crs/rules/REQUEST-942-APPLICATION-ATTACK-
SQLI.conf"] [line "677"] [id "942200"] [msg "Detects MySQL comment-/space-obfuscated injections and backtick termination"] 
[data "Matched Data: , \\x22name\\x22:  found within ARGS:value: {\\x22service\\x22: \\x22SomeService\\x22, \\x22name\\x22: 
\\x22foo\\x22, \\x22doSomething\\x22: false, \\x22CustomerNumber\\x22: \\x22123456789\\x22}"] [severity "CRITICAL"] [ver 
"OWASP_CRS/3.4.0-dev"] [tag "modsecurity"] [tag "application-multi"] [tag "language-multi"] [tag "platform-multi"] [tag "attack-
sqli"] [tag "OWASP_CRS"] [tag "capec/1000/152/248/66"] [tag "PCI/6.5.2"] [tag "paranoia-level/2"] [hostname "localhost"] [uri 
"/index.html"] [unique_id "Yj4jYsdwNgnsGHe769ajlQAAAAA"]
```


```
[Fri Mar 25 20:17:38.349896 2022] [:error] [pid 15:tid 140482004907776] [client 172.x.x.x:58974] [client 172.x.x.x] 
ModSecurity: Warning. Pattern match "(?i)(?:[\\"'`]\\\\s*?(?:(?:n(?:and|ot)|(?:x?x)?or|between|\\\\|\\\\||and|div|&&)\\\\s+[\\\\s\\\\w]+=
\\\\s*?\\\\w+\\\\s*?having\\\\s+|like(?:\\\\s+[\\\\s\\\\w]+=\\\\s*?\\\\w+\\\\s*?having\\\\s+|\\\\W*?[\\"'`\\\\d])|[^?\\\\w\\\\s=.,;)(]++\\\\s*?
[(@\\"'`]*?\\\\s*?\\\\w+\\\\W+\\\\w|\\\\*\\\\s*?\\\\w+\\\\W+[\\"'`])|(?:u ..." at ARGS:value. [file "/etc/modsecurity.d/owasp-crs/rules
/REQUEST-942-APPLICATION-ATTACK-SQLI.conf"] [line "732"] [id "942260"] [msg "Detects basic SQL authentication bypass 
attempts 2/3"] [data "Matched Data: \\x22: \\x22SomeService\\x22, \\x22n found within ARGS:value: {\\x22service\\x22: 
\\x22SomeService\\x22, \\x22name\\x22: \\x22foo\\x22, \\x22doSomething\\x22: false, \\x22CustomerNumber\\x22: 
\\x22123456789\\x22}"] [severity "CRITICAL"] [ver "OWASP_CRS/3.4.0-dev"] [tag "modsecurity"] [tag "application-multi"] [tag 
"language-multi"] [tag "platform-multi"] [tag "attack-sqli"] [tag "OWASP_CRS"] [tag "capec/1000/152/248/66"] [tag "PCI/6.5.2"] 
[tag "paranoia-level/2"] [hostname "localhost"] [uri "/index.html"] [unique_id "Yj4jYsdwNgnsGHe769ajlQAAAAA"]
```


```
[Fri Mar 25 20:17:38.350147 2022] [:error] [pid 15:tid 140482004907776] [client 172.x.x.x:58974] [client 172.x.x.x] 
ModSecurity: Warning. Pattern match "((?:[~!@#\\\\$%\\\\^&\\\\*\\\\(\\\\)\\\\-\\\\+=\\\\{\\\\}\\\\[\\\\]\\\\|:;\\"'\\xc2\\xb4\\xe2\\x80\\x99
\\xe2\\x80\\x98`<>][^~!@#\\\\$%\\\\^&\\\\*\\\\(\\\\)\\\\-\\\\+=\\\\{\\\\}\\\\[\\\\]\\\\|:;\\"'\\xc2\\xb4\\xe2\\x80\\x99\\xe2\\x80\\x98`<>]*?){12})" 
at ARGS:value. [file "/etc/modsecurity.d/owasp-crs/rules/REQUEST-942-APPLICATION-ATTACK-SQLI.conf"] [line "1147"] [id 
"942430"] [msg "Restricted SQL Character Anomaly Detection (args): # of special characters exceeded (12)"] [data "Matched 
Data: {\\x22service\\x22: \\x22SomeService\\x22, \\x22name\\x22: \\x22foo\\x22, \\x22 found within ARGS:value: {\\x22service
\\x22: \\x22SomeService\\x22, \\x22name\\x22: \\x22foo\\x22, \\x22doSomething\\x22: false, \\x22CustomerNumber\\x22: 
\\x22123456789\\x22}"] [severity "WARNING"] [ver "OWASP_CRS/3.4.0-dev"] [tag "modsecurity"] [tag "application-multi"] [tag 
"language-multi"] [tag "platform-multi"] [tag "attack-sqli"] [tag "OWASP_CRS"] [tag "capec/1000/152/248/66"] [tag "PCI/6.5.2"]
 [tag "paranoia-level/2"] [hostname "localhost"] [uri "/index.html"] [unique_id "Yj4jYsdwNgnsGHe769ajlQAAAAA"]
```


We see that the `ARGS:value` (JSON) inside the JSON message triggers the rules.
However, I don't know how common this is and if we can solve it. I will add this issue to the next issue chat meeting.

**Comment by fzipi** [comment]
Honestly, is this a FP?

And I would say it is fixable.... on the client side 😄

**Comment by mirkodziadzka-avi** [comment]
> Honestly, is this a FP?

Per definition of an FP, yes.
 
> And I would say it is fixable.... on the client side 😄

I would not have opened this if it would be a single application. But I saw this the second time now in a different application. So it seems to be "a thing". I will try to find out if this is a common framework or just a way application developers are doing $stuff for $reasons.

I wanted to pick your brains here if there is anything which can be done, except a locally removeTargetById

In an ideal world, I could configure: "ARGS:value" is double encoded JSON, so please parse it and insert it into ARGS with the prefix 'ARGS:value'. Effectively treating it as a big single JSON document. But this seems to be not doable in modsec.

**Comment by dune73** [comment]
That would be elegant indeed. So much useful JSON / XML use cases not supported by ModSec.
